### PR TITLE
ENH: pandas read_* wildcard #15904

### LINKED ISSUE #15904: ENH: pandas read_* wildcard
hi every one,i want to use in pandas like this
``df = pd.read_excel('*order list*.xlsx')``

could panda  wildcard？
thank you！

### PR DESCRIPTION
**Pull request by dwkenefick**
- [x] closes #15904
 - [x] tests added / passed (N/A for docs)
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry (N/A for docs)

### CODE REVIEW

File: doc/source/cookbook.rst [hunk 1]
```diff
@@ -943,6 +940,47 @@ using that handle to read.
...
+        frames.append(pd.read_csv(f))
+    result = pd.concat(frames)
+
+You can use the same approach to read all files matching a pattern.  Here is an example using ``glob``:
+
+.. ipython:: python
+
+    import glob
+    frames = []
+    for f in glob.glob('file_*.csv'):
```
Comment (jreback) [inline doc/source/cookbook.rst]: idiomatically

``result = pd.concat([pd.read_csv(f) for f in glob.glob('file_*.csv')], ignore_index=True)``

usually you want the ``ignore_index=True``

File: doc/source/cookbook.rst [hunk 2]
```diff
@@ -943,6 +940,47 @@ using that handle to read.
...
+       frames.append(pd.read_csv(f))
+    result = pd.concat(frames)
+
+This performs significantly better than using ``pd.append`` to add each of the files to an existing DataFrame.
+Finally, this strategy will work with the other ``read_`` functions described in the :ref:`io docs<io>`.
```
Comment (jreback) [inline doc/source/cookbook.rst]: -> ``pd.read_*(..)``

File: doc/source/cookbook.rst [hunk 3]
```diff
@@ -943,6 +940,47 @@ using that handle to read.
...
+
+.. ipython:: python
+
+    import glob
+    frames = []
+    for f in glob.glob('file_*.csv'):
+       frames.append(pd.read_csv(f))
+    result = pd.concat(frames)
+
+This performs significantly better than using ``pd.append`` to add each of the files to an existing DataFrame.
```
Comment (jreback) [inline doc/source/cookbook.rst]: ``pd.append`` is not a thing. Remove the first sentence in any event.

File: doc/source/cookbook.rst [hunk 4]
```diff
@@ -943,6 +940,47 @@ using that handle to read.
...
+
+.. ipython:: python
+
+    for i in range(3):
+        data = pd.DataFrame(np.random.randn(10, 4))
+        data.to_csv('file_{}.csv'.format(i))
+
+    frames = []
+    files = ['file_0.csv', 'file_1.csv', 'file_2.csv']
+    for f in files:
```
Comment (jreback) [inline doc/source/cookbook.rst]: see below for the idiom to do this

File: doc/source/cookbook.rst [hunk 5]
```diff
@@ -943,6 +940,47 @@ using that handle to read.
 `Write a multi-row index CSV without writing duplicates
 <http://stackoverflow.com/questions/17349574/pandas-write-multiindex-rows-with-to-csv>`__
 
+.. _cookbook.csv.multiple_files:
+
+Reading multiple files to create a single DataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The best way to combine multiple files into a single DataFrame is to read the individual frames one by one, put all
+of the individual frames into a list, and then combine the frames in the list using ``pd.concat``:
```
Comment (jreback) [inline doc/source/cookbook.rst]: use ` `:func:pd.concat` ``

File: doc/source/cookbook.rst (hunk 1 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: added

File: doc/source/cookbook.rst (hunk 3 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: Removed here and in io.rst

File: doc/source/cookbook.rst (hunk 5 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: Changed here and in io.rst

File: doc/source/cookbook.rst (hunk 2 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: added, but with `(...)` to be consistent with a later usage.

File: doc/source/cookbook.rst [hunk 6]
```diff
@@ -943,6 +940,43 @@ using that handle to read.
...
+The best way to combine multiple files into a single DataFrame is to read the individual frames one by one, put all
+of the individual frames into a list, and then combine the frames in the list using :func:`pd.concat`:
+
+.. ipython:: python
+
+    for i in range(3):
+        data = pd.DataFrame(np.random.randn(10, 4))
+        data.to_csv('file_{}.csv'.format(i))
+
+    frames = []
```
Comment (TomAugspurger) [inline doc/source/cookbook.rst]: You can delete this `frames` now.

File: doc/source/cookbook.rst [hunk 7]
```diff
@@ -943,6 +940,43 @@ using that handle to read.
...
+    files = ['file_0.csv', 'file_1.csv', 'file_2.csv']
+    result = pd.concat([pd.read_csv(f) for f in files], ignore_index=True)
+
+You can use the same approach to read all files matching a pattern.  Here is an example using ``glob``:
+
+.. ipython:: python
+
+    import glob
+    frames = []
```
Comment (TomAugspurger) [inline doc/source/cookbook.rst]: Same with this `frames`

**Comment by jreback** [comment]
I think this should have a pointer from the ``io.rst/read_csv`` docs as well (maybe a mini section), or it *could* entirely exist there (and have a pointer from the cookbook)

File: doc/source/cookbook.rst (hunk 6 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: Removed, thanks.

File: doc/source/cookbook.rst (hunk 7 above)
Comment (dwkenefick) [inline doc/source/cookbook.rst]: Removed

**Comment by dwkenefick** [comment]
@jreback There should be a small section in `io.rst` that references this.  See [here](https://github.com/pandas-dev/pandas/pull/16166/files#diff-bb262b5e18f651a4357476b250d4b255R1448).  If you had something different in mind let me know - happy to make the change.

**Comment by TomAugspurger** [comment]
@dwkenefick thanks! The doc build should be done in 20-30 minutes, if you want to check the output [here](http://pandas-docs.github.io/pandas-docs-travis/cookbook.html)

**Pull request merged**